src/Test/TestCombatMods.java

```diff
@@ -0,0 +1,131 @@
+    // ...and defense
+    BattleSummary vengefulCounter = CombatEngine.simulateBattleResults(infActive, meaty, testMap, 8, 3);
+    BattleSummary normalCounter = CombatEngine.simulateBattleResults(infPassive, meaty, testMap, 8, 5);
+    
+    // Check that Venge's passive ability works on both attack and defense
+    boolean testPassed = validate(vengeful.defenderHPLoss > normal.defenderHPLoss, "    Being angry didn't help Venge attack extra hard.");
+    testPassed &= validate(vengefulCounter.attackerHPLoss > normalCounter.attackerHPLoss, "    Being angry didn't help Venge defend extra hard.");
+
+    // Clean up
+    testMap.removeUnit(infActive);
+    testMap.removeUnit(infPassive);
+    testMap.removeUnit(bait);
+    testMap.removeUnit(meaty);
+
+    return testPassed;
+  }
+
+  /** Test that combat works as a black box, but that Venge gets his way all the same. */
+  private boolean testRetribution()
+  {
+    // Add our test subjects
+    Unit infA = addUnit(testMap, cinder, UnitEnum.INFANTRY, 7, 3);
+    Unit infB = addUnit(testMap, venge, UnitEnum.INFANTRY, 7, 5);
+    
+    // Check our damage for each first strike pre-power...
+    BattleSummary normalAB = CombatEngine.simulateBattleResults(infA, infB, testMap, 7, 4);
+    BattleSummary normalBA = CombatEngine.simulateBattleResults(infB, infA, testMap, 7, 4);
+
+    venge.modifyAbilityPower(42); // juice up
+    venge.getReadyAbilities().get(1).activate(testMap); // activate Retribution
+    
+    // ...and after power
+    BattleSummary retribAB = CombatEngine.simulateBattleResults(infA, infB, testMap, 7, 4);
+    BattleSummary retribBA = CombatEngine.simulateBattleResults(infB, infA, testMap, 7, 4);
+    
+    // Check that Venge's Retribution works properly without breaking things (other than balance)
+    boolean testPassed = true;
+    
+    // First, check the logic of A->B
+    testPassed &= validate(normalAB.defenderHPLoss > normalAB.attackerHPLoss, "    First strike didn't work properly for Cinder.");
+
+    testPassed &= validate(retribAB.attacker == infA, "    infA attacked, but isn't the attacker.");
+    testPassed &= validate(retribAB.defender == infB, "    infB was attacked, but isn't the defender.");
+    testPassed &= validate(retribAB.defenderHPLoss < retribAB.attackerHPLoss, "    Cinder got first strike when Retribution should have stolen it.");
+
+    // Now do B->A
+    testPassed &= validate(normalBA.defenderHPLoss > normalBA.attackerHPLoss, "    First strike didn't work properly for Venge.");
+
+    testPassed &= validate(retribBA.attacker == infB, "    infB attacked, but isn't the attacker.");
+    testPassed &= validate(retribBA.defender == infA, "    infA was attacked, but isn't the defender.");
+    testPassed &= validate(retribBA.defenderHPLoss > retribBA.attackerHPLoss, "    Retribution somehow deprived Venge of first strike?");
+
+    // Clean up
+    testMap.removeUnit(infA);
+    testMap.removeUnit(infB);
+
+    return testPassed;
```

Comment:
Should we also test that Venge's defenses are lower with Retribution active, and that firepower is actually higher? Maybe those would be better as tests for the two modifier classes specifically, but verifying they are correctly used in the ability here doesn't hurt either.

What about a test for IronWill, since it's mucking with counter-attack damage?

I lean towards thorough testing when I can, but I know overkill is also a thing, so add what you think makes sense.

Reply:
I wouldn't be able to test the defense combat-wise without assuming I know all the internals of the damage formula. If that ever changes, it would also be annoying to update this.
I could build my own BattleParams, but that feels like a little much.

I can check the stats and Venge's attack damage, though.